### PR TITLE
Feature - Populate Tamara Order

### PR DESCRIPTION
**Pull request by loikimle**

### CODE REVIEW

File: src/App/Services/Tamara_Client.php [hunk 1]
```diff
@@ -54,6 +75,467 @@ public function reinit_tamara_client( $api_token, $api_url = 'https://api.tamara
...
 
+	/**
+	 * Create Tamara Checkout session, if failed, put errors to `wc_notices`
+	 *
+	 * @param $wc_order_id
+	 *
+	 * @return array|bool
+	 * @throws \Exception
+	 */
+	public function tamara_checkout_session( $wc_order_id ) {
```
Comment (npbtrac) [inline src/App/Services/Tamara_Client.php:86]: you should rename to `create_checkout_session()`

File: src/App/Services/Tamara_Client.php [hunk 2]
```diff
@@ -54,6 +75,467 @@ public function reinit_tamara_client( $api_token, $api_url = 'https://api.tamara
...
+		$wc_order = wc_get_order( $wc_order_id );
+		$instalment_period = 3;
+		$checkout_payment_type = 'PAY_BY_INSTALMENTS';
+		try {
+			$checkout_response = $this->create_tamara_checkout_session(
```
Comment (npbtrac) [inline src/App/Services/Tamara_Client.php:91]: we need to use the variable name `$create_checkout_session_response`

File: src/App/Services/Tamara_Client.php [hunk 3]
```diff
@@ -54,6 +75,467 @@ public function reinit_tamara_client( $api_token, $api_url = 'https://api.tamara
...
+				$wc_order,
+				$checkout_payment_type,
+				$instalment_period
+			);
+		} catch ( Exception $tamaraCheckoutException ) {
```
Comment (npbtrac) [inline src/App/Services/Tamara_Client.php:96]: Recheck the case

File: src/App/Services/Tamara_Client.php [hunk 4]
```diff
@@ -54,6 +75,467 @@ public function reinit_tamara_client( $api_token, $api_url = 'https://api.tamara
...
+			$errorMessage = $this->tamara_checkout_wp_plugin->_t( 'Tamara Service unavailable! Please try again later.' );
```
Comment (npbtrac) [inline src/App/Services/Tamara_Client.php:97]: You should create a method for this `_t` and on that method, you use `Tamara_Checkout_WP_Plugin::wp_app_instance()->_()`

File: src/App/Services/Tamara_Client.php [hunk 5]
```diff
@@ -54,6 +75,467 @@ public function reinit_tamara_client( $api_token, $api_url = 'https://api.tamara
...
+	 *
+	 * @param  WC_Order  $wc_order
+	 *
+	 * @param $payment_type
+	 * @param $instalment_period
+	 *
+	 * @return CreateCheckoutResponse
+	 * @throws Exception
+	 */
+	public function create_tamara_checkout_session( WC_Order $wc_order, $payment_type, $instalment_period ): CreateCheckoutResponse {
```
Comment (npbtrac) [inline src/App/Services/Tamara_Client.php:140]: we may rename this to `build_checkout_session_request()` as we are in the Tamara_Client so we don't need to refer to tamara again

File: src/App/Services/Tamara_Client.php [hunk 6]
```diff
@@ -54,6 +75,467 @@ public function reinit_tamara_client( $api_token, $api_url = 'https://api.tamara
...
+		$client = $this->api_client;
+		$checkoutRequest = new CreateCheckoutRequest(
```
Comment (npbtrac) [inline src/App/Services/Tamara_Client.php:142]: check the case

File: src/App/Support/Helpers/MoneyHelper.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Tamara_Checkout\App\Support\Helpers;
+
+class MoneyHelper {
```
Comment (npbtrac) [inline src/App/Support/Helpers/MoneyHelper.php:5]: Recheck the case

File: src/App/Support/Traits/Tamara_Order_Trait.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tamara_Checkout\App\Support\Traits;
+
+trait Tamara_Order_Trait {
```
Comment (npbtrac) [inline src/App/Support/Traits/Tamara_Order_Trait.php:7]: Review the name, the trait should have a specific name for it's purpose, e.g. `Tamara_Order_Error_Mappings_Trait` ...

File: src/App/Support/Traits/Wc_Order_Settings_Trait.php [hunk 9]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tamara_Checkout\App\Support\Traits;
+
+trait Wc_Order_Settings_Trait {
```
Comment (npbtrac) [inline src/App/Support/Traits/Wc_Order_Settings_Trait.php:7]: Review the name like above comment

File: src/App/Support/Traits/Wc_Order_Settings_Trait.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+			? $this->get_currency_to_country_mapping()[ get_woocommerce_currency() ]
+			: $this->get_store_base_country_code();
+	}
+
+	/**
+	 * Get country code based on its currency
+	 *
+	 * @return array
+	 */
+	public function get_currency_to_country_mapping(): array {
```
Comment (npbtrac) [inline src/App/Support/Traits/Wc_Order_Settings_Trait.php:20]: We need to put to a Helper class as it's widely used

File: src/App/Support/Traits/Wc_Order_Settings_Trait.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+			'BHD' => 'BH',
+		];
+	}
+
+	/**
+	 * Get store base country code
+	 *
+	 * @return string
+	 */
+	public function get_store_base_country_code(): string {
```
Comment (npbtrac) [inline src/App/Support/Traits/Wc_Order_Settings_Trait.php:34]: Consider to move this to a Helper as well, there's a method in Tamara_Widget, you may need to refactor it as well

File: src/App/Support/Traits/Wc_Order_Settings_Trait.php [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	/**
+	 * Remove trailing slashes of an url
+	 *
+	 * @param  string  $url
+	 *
+	 * @return string
+	 */
+	public function remove_trailing_slashes( string $url ): string {
```
Comment (npbtrac) [inline src/App/Support/Traits/Wc_Order_Settings_Trait.php:47]: Consider to move to a helper class.

File: src/App/Jobs/Register_Tamara_WP_Api_Routes_Job.php [hunk 13]
```diff
@@ -14,7 +14,7 @@ class Register_Tamara_WP_Api_Routes_Job extends Base_Job {
 
 	public function handle(): void {
 		Route::post( 'tamara/success', [ Main_Controller::class, 'handle_tamara_success' ] )->name( 'tamara-success' );
-		Route::post( 'tamara/cancel', [ Main_Controller::class, 'handle_tamara_cancel' ] )->name( 'tamara-cancel' );
+		Route::get( 'tamara/orders/{wc_order_id}/cancel', [ Main_Controller::class, 'handle_tamara_cancel' ] )->name( 'wp-api-tamara-cancel' );
```
Comment (tracnguyenp) [inline src/App/Jobs/Register_Tamara_WP_Api_Routes_Job.php:17]: @loikimle you can check this commit to see how we use the routes. For General Helper, you should use static method for all and you don't need create an object to that class.

You don't need to use $this->tamara_checkout_wp_plugin, just use `Tamara_Checkout_WP_Plugin::wp_app_instance()`

**Pull request merged**